src/coreclr/jit/emitarm64.cpp

```diff
@@ -7987,18 +8002,7 @@ void emitter::emitIns_R_I(instruction     ins,
     assert(canEncode);
     assert(fmt != IF_NONE);
 
-    instrDesc* id;
-
-    if (!optionalShift)
-    {
-        id = emitNewInstrSC(attr, imm);
-    }
-    else
-    {
-        // Instructions with optional shifts (MOV, DUP, etc.) need larger instrDesc to store state
-        id = emitNewInstrCns(attr, imm);
-        id->idOptionalShift(hasShift);
-    }
+    instrDesc* id = emitNewInstrSC(attr, imm);
```

Comment:
Looks like the `imm` isn't being converted before calling `emitNewInstrSC()`. The larger MultipleOf256 immediates are not going to fit into a normal instr?

Reply:
yes, i think I noticed that too.

Reply:
Yeah, that's probably not ideal... Here's my new approach:
* For encodings without any shifting, just call `emitNewInstrSC`.
* For unshifted immediates, call `emitNewInstrSC`. If it allocates a small descriptor, `idHasShift` will always return false. Else, `idHasShift` still returns false, as we set the dedicated bit in large descriptors to false after creating the descriptor.
* For immediates that need a shift, call `emitNewInstrCns` so a normal or large descriptor is used. `idHasShift` will always check the dedicated bit, as it is always available. We set this bit to true after creating the descriptor.

This means we can use small descriptors more often than we could with the prior approach of always calling `emitNewInstrCns` for encodings that might have shifts. The downside to this approach is `idHasShift` has to also check if the descriptor is small... If we want to prioritize TP over memory usage, then the previous approach of not converting `imm` before saving it to the descriptor might be better.

Reply:
> prioritize TP over memory usage

Yes, we need to be clear what we're optimising for, and that'll dictate when we should be encoding/decoding.  I'll leave it to others to decide which approach we should be going for.

Reply:
I am ok with this approach.

Reply:
> If we want to prioritize TP over memory usage

For SVE encodings, which I expect to be very rare, I don't think we need to worry about either TP or memory. I would prioritize understandable code first, TP next.

I would be very highly be concerned with TP/memory impact on non-SVE encodings.

Reply:
The previous approach of allocating large descriptors for shifted values sacrificed memory usage for SVE encodings in favor of improving TP for non-SVE encodings by removing a branch or two in the affected `emitIns_*` methods. The diffs showed that only improved TP by <0.1% though, so I'm not too attached to either approach.

> I would prioritize understandable code first, TP next.

I think the previous approach is easier to understand since there's less state to track, though that approach requires us to encode the immediate (i.e. right-shift it by 8) each time we print it, encode it in an instruction, etc.